Apply monkey patches for SQL Server connections only

mattbrictson opened the pull request:
Recent versions of Rails support multiple database connections within the same app. It is possible for these connections to use different adapters. For example, one adapter may use SQL Server, and another uses PostgreSQL.

This gem applies some monkey patches to ActiveRecord for SQL Server compatibility. These patches could break other adapters, though, in a multiple-database scenario.

This PR modifies the patches so that they are applied only if the connection is SQL Server. If not, the original ActiveRecord implementation (`super`) is used instead.

Fixes #929

@aidanharan @wpolicarpo what do you think of this approach? Any suggestions on how to test?

mattbrictson commented:
Hmm. Appveyor CI is failing with [one failing test](https://ci.appveyor.com/project/rails-sqlserver/activerecord-sqlserver-adapter/builds/40053460/job/n5phge5t2bii31r3#L369), but that same test fails for me locally on `main`...

On `main`, running locally:

```
Failure:
LogSubscriberTest#test_where_in_binds_logging_include_attribute_names [/Users/mattb/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/bundler/gems/rails-f94f41dd59b5/activerecord/test/cases/log_subscriber_test.rb:252]:
Expected /\["id",\ 1\],\ \["id",\ 2\],\ \["id",\ 3\],\ \["id",\ 4\],\ \["id",\ 5\]/ to match "Developer Load (2.3ms)  EXEC sp_executesql N'SELECT [developers].[id], [developers].[name], [developers].[salary], [developers].[firm_id], [developers].[mentor_id], [developers].[legacy_created_at], [developers].[legacy_updated_at], [developers].[legacy_created_on], [developers].[legacy_updated_on] FROM [developers] WHERE [developers].[id] IN (@0, @1, @2, @3, @4)', N'@0 bigint, @1 bigint, @2 bigint, @3 bigint, @4 bigint', @0 = 1, @1 = 2, @2 = 3, @3 = 4, @4 = 5  [[\"id\", nil], [\"id\", nil], [\"id\", nil], [\"id\", nil], [\"id\", nil]]".
```

aidanharan commented:
Yeah this fix looks fine to me. The `LogSubscriberTest#test_where_in_binds_logging_include_attribute_names` test should be fixed already by https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/931

wpolicarpo commented:
Also, could you rebase your branch and add a changelog entry?

mattbrictson commented:
@wpolicarpo thanks for the feedback! I think I've addressed all of your concerns:

- Rebased against main
- Changed `is_a?(SQLServerAdapter)` to `adapter_name == "SQLServer"`
- Added an **Unreleased** section to the CHANGELOG with an entry for this PR

I also squashed this PR down to one commit.

Let me know if there is anything else you need.

wpolicarpo commented:
All good. Thanks for your contribution.
